clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py

```diff
@@ -0,0 +1,586 @@
+    complete_data_df = complete_data_df.set_index(
+        ["participant_id", "session_id", "modality", "bids_filename"],
+        verify_integrity=True,
+    )
+    # open the reference for building the tsvs:
+    path_to_ref_csv = os.path.join(
+        os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
+        "data",
+        "genfi_ref.csv",
+    )
+    df_ref = pd.read_csv(path_to_ref_csv, sep=";")
+    if not gif:
+        print(df_ref.head(8))
+    return {
+        col: complete_data_df.filter(items=list(df_ref[col]))
+        for col in ["participants", "sessions", "scans"]
+    }
+
+
+def intersect_data(
+    imaging_data: DataFrame, df_clinical_complete: DataFrame
+) -> DataFrame:
+    """This function merges the dataframe containing the data extracted from the raw images and from the clinical data
+
+    Parameters
+    ----------
+    imaging_data: DataFrame
+        Dataframe containing the data extracted from the raw images
+
+    df_clinical_complete: DataFrame
+        Dataframe containing the clinical data
+
+    Returns
+    -------
+    df_complete: DataFrame
+        Dataframe containing the merged data
+    """
+    df_complete = imaging_data.merge(
+        df_clinical_complete,
+        how="inner",
+        left_on=["source_id", "source_ses_id"],
+        right_on=["blinded_code", "visit"],
+    )
+    df_complete = df_complete.loc[:, ~df_complete.columns.duplicated()]
+    return df_complete
+
+
+def read_imaging_data(source_path: PathLike) -> DataFrame:
+    """This function finds the imaging data and filters it.
+
+    Parameters
+    ----------
+    source_path: PathLike
+        Path to the raw data
+
+    Returns
+    -------
+    df_dicom: DataFrame
+        Dataframe containing the data extracted.
+    """
+    df_dicom = pd.DataFrame(find_dicoms(source_path), columns=["source_path", "source"])
+    df_dicom = filter_dicoms(df_dicom)
+    return df_dicom
+
+
+def complete_imaging_data(df_dicom: DataFrame) -> DataFrame:
+    """This function uses the raw information extracted from the images, to obtain all the information necessary for the BIDS
+
+    Parameters
+    ----------
+    df_dicom: DataFrame
+        Dataframe containing the data extracted from the images
+
+    Returns
+    -------
+    df_sub_ses_run: DataFrame
+        Dataframe with the data necessary for the BIDS
+    """
+    from clinica.iotools.bids_utils import identify_modality
+
+    df_dicom = df_dicom.assign(
+        source_id=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(x).parent).parent).name.split("-")[0]
+        )
+    )
+    df_dicom = df_dicom.assign(
+        source_ses_id=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(x).parent).parent).name.split("-")[1]
+        )
+    )
+    df_dicom = df_dicom.assign(
+        origin=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(Path(Path(x).parent).parent).parent).parent)
+        )
+    )
```

Comment:
I'm starting the review of this beast.
For the first half of the function, I think we could perform the following extractions:

```python
def compute_baseline_date(df: pd.DatFrame) -> pd.DataFrame:
    """Computes the baseline date by taking the minimum
    acq_date for each subject.
    """
    df_1 = (
         df[["source_id", "source_ses_id", "acq_date"]]
         .groupby(["source_id", "source_ses_id"])
         .min()
     )
     df_2 = df[["source_id", "acq_date"]].groupby(["source_id"]).min()
     return df_1.join(df_2.rename(columns={"acq_date": "baseline"}))

def compute_session_numbers(df: pd.DatFrame) -> pd.DataFrame:
    """Computes the session number in months by subtracting the
    baseline date from the acq_date and converting to months.
    """
     df = df.assign(
         ses_month=lambda x: (
             (
                 x["acq_date"].str[:4].astype("int")
                 - x["baseline"].str[:4].astype("int")
             )
             * 12
             + (
                 x["acq_date"].str[4:6].astype("int")
                 - x["baseline"].str[4:6].astype("int")
             )
         )
     )
     return df.assign(
         session_id=lambda x: x.ses_month.map(lambda y: f"ses-M{y:03d}")
     )

def compute_modality(df: pd.DataFrame) -> pd.DataFrame:
    """Parses the series_desc column to identify the modality, and map
    this modality to a metadata dictionary.
    """
    from clinica.iotools.bids_utils import identify_modality
    
    modality_mapping = {
                 "T2w": {
                     "datatype": "anat",
                     "suffix": "T2w",
                     "sidecars": ["T2w.json"],
                     "task": "",
                 },
                 "T1": {
                     "datatype": "anat",
                     "suffix": "T1w",
                     "sidecars": ["T1w.json"],
                     "task": "",
                 },
                 "dwi": {
                     "datatype": "dwi",
                     "suffix": "dwi",
                     "sidecars": [
                         "dwi.json",
                         "dwi.bval",
                         "dwi.bvec",
                     ],
                     "task": "",
                 },
                 "rsfmri": {
                     "datatype": "func",
                     "suffix": "bold",
                     "sidecars": ["rsfmri.json"],
                     "task": "_task-rest",
                 },
                 "fieldmap": {
                     "datatype": "fmap",
                     "suffix": "fmap",
                     "sidecars": ["fmap.json"],
                     "task": "",
                 },
                 # "asl": {
                 #     "datatype": "perf",
                 #     "suffix": "asl",
                 #     "sidecars": ["asl.json"],
                 #     "task": "",
                 # },
     }
    df = df.assign(
         modality=lambda x: x.series_desc.apply(lambda y: identify_modality(y))
     )
     return df.join(
         df.modality.map(modality_mapping).apply(pd.Series)
     )
```

which would then give us :

```python
(...)
df_1 = compute_baseline_date(df_dicom)
df_1 = compute_session_numbers(df_1)
df_sub_ses = df_dicom.merge(
     df_1[["baseline", "session_id"]], how="inner", on=["source_id", "source_ses_id"]
 )
 df_sub_ses = df_sub_ses.assign(
     participant_id=lambda df: df.source_id.apply(lambda x: f"sub-{x}")
 )
df_sub_ses = compute_modality(df_sub_ses)
(...)
```

Does this make sense ?